lib/vrf.c

```diff
@@ -994,25 +994,50 @@ const char *vrf_get_default_name(void)
 	return vrf_default_name;
 }
 
-int vrf_bind(vrf_id_t vrf_id, int fd, const char *name)
+int vrf_bind(vrf_id_t vrf_id, int fd, const char *ifname)
 {
 	int ret = 0;
 	struct interface *ifp;
+	struct vrf *vrf;
 
-	if (fd < 0 || name == NULL)
+	if (fd < 0)
 		return fd;
-	/* the device should exist
-	 * otherwise we should return
-	 * case ifname = vrf in netns mode => return
-	 */
-	ifp = if_lookup_by_name(name, vrf_id);
-	if (!ifp)
+
+	if (vrf_id == VRF_UNKNOWN)
 		return fd;
+
+	/* can't bind to a VRF that doesn't exist */
+	vrf = vrf_lookup_by_id(vrf_id);
+	if (!vrf_is_enabled(vrf))
+		return -1;
+
+	if (ifname && strcmp(ifname, vrf->name)) {
```

Comment:
I don't think I get this question and its relevance to this PR.

Reply:
In this code, when ifname is interface (say Ethernet64), we are processing it differently (by calling if_lookup_by_name() and checking inside if_name_head database). But if ifname is vrf (e.g. Vrf-Red), we will not check inside the database. I was wondering why we have this difference in code. From zebra perspective, both look same as below.
sonic# show interface Vrf1
Interface Vrf1 is up, line protocol is up
  Link ups:       0    last: (never)
  Link downs:     0    last: (never)
  vrf: Vrf1
  index 8 metric 0 mtu 65536 speed 0 
  flags: <UP,RUNNING,NOARP>
  group: 0
  Type: Ethernet
  HWaddr: a6:2b:4f:6a:2a:b5
  Interface Type VRF
sonic# show interface Ethernet64
Interface Ethernet64 is up, line protocol is up
  Link ups:       0    last: (never)
  Link downs:     0    last: (never)
  vrf: default
  index 64 metric 0 mtu 9100 speed 100000 
  flags: <UP,BROADCAST,RUNNING,MULTICAST>
  group: 0
  Type: Ethernet
  HWaddr: b8:6a:97:e2:5e:9f
  inet 64.1.1.1/24
  Interface Type Other

Reply:
This is done, because the code flow in the daemons is following
1. receive vrf registration from zebra
2. create and bind socket to the vrf
3. receive interface registration from zebra

If we try to look for an interface on the second step, we always fail, because we didn't receive the interface registration message yet. It's enough to check for the VRF existence if the ifname is the VRF name.